backend/chunker/chunker.go

```diff
@@ -0,0 +1,1412 @@
+			return fmt.Errorf("%q misses chunks", o)
+		}
+		totalSize += chunk.Size()
+	}
+	o.size = totalSize
+	return nil
+}
+
+func (f *Fs) newObject(remote string, main fs.Object, chunks []fs.Object) *Object {
+	var size int64 = -1
+	if main != nil {
+		size = main.Size()
+		if remote == "" {
+			remote = main.Remote()
+		}
+	}
+	return &Object{
+		remote: remote,
+		main:   main,
+		size:   size,
+		f:      f,
+		chunks: chunks,
+	}
+}
+
+// Fs returns read only access to the Fs that this object is part of
+func (o *Object) Fs() fs.Info {
+	return o.f
+}
+
+// Return a string version
+func (o *Object) String() string {
+	if o == nil {
+		return "<nil>"
+	}
+	return o.remote
+}
+
+// Remote returns the remote path
+func (o *Object) Remote() string {
+	return o.remote
+}
+
+// Size returns the size of the file
+func (o *Object) Size() int64 {
+	if o.chunks != nil {
+		return o.size
+	}
+	return o.main.Size()
+}
+
+// Storable returns whether object is storable
+func (o *Object) Storable() bool {
+	if o.main != nil {
+		return o.main.Storable()
+	}
+	if o.chunks == nil {
+		panic("invalid chunked object")
+	}
+	return o.chunks[0].Storable()
+}
+
+// ModTime returns the modification time of the file
+func (o *Object) ModTime(ctx context.Context) time.Time {
+	if o.main != nil {
+		return o.main.ModTime(ctx)
+	}
+	if o.chunks == nil {
+		panic("invalid chunked object")
+	}
+	return o.chunks[0].ModTime(ctx)
+}
+
+// SetModTime sets the modification time of the file
+func (o *Object) SetModTime(ctx context.Context, mtime time.Time) error {
+	if o.main != nil {
+		return o.main.SetModTime(ctx, mtime)
+	}
+	if o.chunks == nil {
+		panic("invalid chunked object")
+	}
+	return o.chunks[0].SetModTime(ctx, mtime)
+}
+
+// Hash returns the selected checksum of the file
+// If no checksum is available it returns ""
+func (o *Object) Hash(ctx context.Context, ht hash.Type) (string, error) {
+	return "", hash.ErrUnsupported
+}
+
+// UnWrap returns the wrapped Object
+func (o *Object) UnWrap() fs.Object {
+	if o.main != nil {
+		return o.main
+	}
+	if o.chunks == nil {
+		panic("invalid chunked object")
+	}
+	return o.chunks[0]
+}
+
+// Open opens the file for read.  Call Close() on the returned io.ReadCloser
+func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (rc io.ReadCloser, err error) {
+	if o.main == nil && o.chunks == nil {
+		panic("invalid chunked object")
+	}
+	if o.chunks == nil {
+		return o.main.Open(ctx, options...)
+	}
+
+	var openOptions []fs.OpenOption
+	var offset, limit int64 = 0, -1
+	for _, option := range options {
+		switch opt := option.(type) {
+		case *fs.SeekOption:
+			offset = opt.Offset
+		case *fs.RangeOption:
+			offset, limit = opt.Decode(o.size)
+		default:
+			// pass on Options to underlying open if appropriate
+			openOptions = append(openOptions, option)
+		}
+	}
+	if limit < 0 {
+		limit = o.size - offset
+	}
+
+	var mrc multiReadCloser
+	var readers []io.Reader
+
+	for _, chunk := range o.chunks {
```

Comment:
I think opening all the chunks in advance will be very inefficient for small chunk sizes?

Wouldn't it be better to just have one chunk open at a time?

Reply:
Right, it's suboptimal. Then I'll need something more intelligent than io.MultiReader(readers...), a custom reader that can open another chunk when previous chunk reading ends, and passes errors appropriately. Doesn't look an easy job. Can we have it a separate PR?

Reply:
In a general case it depends on what is optimized: time to first byte or a total download time (it may be favorable to start parallel downloads asap).

Reply:
implemented in the next round of commits